tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1013 @@
+	currentSegmentID   uint32
+	lastFileID         uint32
+	totalSize          uint64
+	lastWrite          time.Time
+	maxSegmentSize     int
+	cache              map[string][][]byte
+	// this cache is a temporary placeholder to keep data while its being flushed
+	// and compacted. It's for cursors to combine the cache and this if a flush is occuring
+	flushCache        map[string][][]byte
+	cacheDirtySort    map[string]bool // will be true if the key needs to be sorted
+	cacheSizes        map[string]int
+	compactionRunning bool
+}
+
+func newPartition(id uint8, path string, segmentSize int) (*partition, error) {
+	return &partition{
+		id:             id,
+		path:           path,
+		maxSegmentSize: segmentSize,
+		lastWrite:      time.Now(),
+		cache:          make(map[string][][]byte),
+		cacheDirtySort: make(map[string]bool),
+		cacheSizes:     make(map[string]int),
+	}, nil
+}
+
+// Close resets the caches and closes the currently open segment file
+func (p *partition) Close() error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.cache = nil
+	p.cacheDirtySort = nil
+	p.cacheSizes = nil
+	if err := p.currentSegmentFile.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// write will write a compressed block of the points to the current segment file. If the segment
+// file is larger than the max size, it will roll over to a new file before performing the write.
+// This method will also add the points to the in memory cache
+func (p *partition) write(points []tsdb.Point) error {
+	block := make([]byte, 0)
+	for _, pp := range points {
+		block = append(block, marshalWALEntry(pp.Key(), pp.UnixNano(), pp.Data())...)
+	}
+	b := snappy.Encode(nil, block)
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// rotate to a new file if we've gone over our limit
+	if p.currentSegmentFile == nil || p.currentSegmentSize > p.maxSegmentSize {
+		err := p.newSegmentFile()
+		if err != nil {
+			return err
+		}
+	}
+
+	if n, err := p.currentSegmentFile.Write(u64tob(uint64(len(b)))); err != nil {
+		return err
+	} else if n != 8 {
+		return fmt.Errorf("expected to write %d bytes but wrote %d", 8, n)
+	}
+
+	if n, err := p.currentSegmentFile.Write(b); err != nil {
+		return err
+	} else if n != len(b) {
+		return fmt.Errorf("expected to write %d bytes but wrote %d", len(b), n)
+	}
+
+	if err := p.currentSegmentFile.Sync(); err != nil {
+		return err
+	}
+
+	p.currentSegmentSize += (8 + len(b))
+	p.lastWrite = time.Now()
+
+	for _, pp := range points {
+		p.addToCache(pp.Key(), pp.Data(), pp.UnixNano())
+	}
+	return nil
+}
+
+// newSegmentFile will close the current segment file and open a new one, updating bookkeeping info on the partition
+func (p *partition) newSegmentFile() error {
+	p.currentSegmentID += 1
+	if p.currentSegmentFile != nil {
+		if err := p.currentSegmentFile.Close(); err != nil {
+			return err
+		}
+	}
+
+	fileName := p.fileNameForSegment(p.currentSegmentID)
+
+	ff, err := os.OpenFile(fileName, os.O_CREATE|os.O_RDWR, 0600)
```

Comment:
Was following what was done in the hinted handoff stuff. Should we be making this same change there as well?